src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,106 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private int size = 0;
+    private Object[] elements;
+
+    public ArrayList() {
+        elements = new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        ensureCapacity();
+        elements[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index");
+        }
+        ensureCapacity();
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getElementAtIndex(index);
     }
 
     @Override
     public void set(T value, int index) {
-
+        T oldValue = getElementAtIndex(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        final T removedValue = getElementAtIndex(index);
+        int numToCopy = size - index - 1;
+        if (numToCopy > 0) {
+            System.arraycopy(elements, index + 1, elements, index, numToCopy);
+        }
+        elements[size - 1] = null;
+        size--;
+        return removedValue;
     }
 
-    @Override
     public T remove(T element) {
-        return null;
+        int index = indexOf(element);
+        if (index != -1) {
+            return remove(index);
+        } else {
+            throw new NoSuchElementException("Element not found " + element);
+        }
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void ensureCapacity() {
+        if (size == elements.length) {
+            int newCapacity = elements.length * 2;
```

Comment:
GROW_FACTOR should be 1.5
```suggestion
            int newCapacity = elements.length * GROW_FACTOR;
```